ctl/src/main.rs

```diff
@@ -0,0 +1,133 @@
+use raft_engine::Error;
+use raft_engine::Result;
+use std::path::PathBuf;
+use structopt::StructOpt;
+
+#[derive(Debug, StructOpt)]
+#[structopt[name="basic"]]
+struct ControlOpt {
+    // sub command type
+    #[structopt[subcommand]]
+    cmd: Option<Cmd>,
+}
+
+#[derive(Debug, StructOpt)]
+pub enum Cmd {
+    /// dump out all operations in log files
+    Dump {
+        /// Path of a specific raft-engine file
+        #[structopt(short, long)]
+        file: Option<String>,
+
+        /// Path of raft-engine storage directory
+        #[structopt(short, long, required_unless = "file")]
+        path: Option<String>,
+
+        /// raft_group ids(optional), format: raft_groups_id1,raft_group_id2....
+        #[structopt(short, long)]
+        raft_groups: Option<String>,
+    },
+
+    /// check log files for logical errors
+    Check {
+        /// Path of directory
+        #[structopt(short, long)]
+        path: Option<String>,
+    },
+
+    /// check log files for logical errors
+    Truncate {
+        /// Path of raft-engine storage directory
+        #[structopt(short, long)]
+        path: Option<String>,
+
+        /// check mode
+        #[structopt(short, long)]
+        mode: Option<String>,
+
+        /// queue name
+        #[structopt(short, long)]
+        queue: Option<String>,
+
+        /// raft_group ids(optional), format: raft_groups_id1,raft_group_id2....
+        #[structopt(short, long)]
+        raft_groups: Option<String>,
+    },
+
+    /// repair log entry holes by fill in empty message
+    Autofill {
+        /// Path of raft-engine storage directory
+        #[structopt(short, long)]
+        path: Option<String>,
+
+        /// queue name
+        #[structopt(short, long)]
+        queue: Option<String>,
+
+        /// raft_group ids(optional), format: raft_groups_id1,raft_group_id2....
+        #[structopt(short, long)]
+        raft_groups: Option<Vec<u32>>,
+    },
+}
+
+impl ControlOpt {
+    pub fn validate_and_execute(&self) -> Result<()> {
+        if self.cmd.is_none() {
+            return Err(Error::InvalidArgument("subcommand is needed".to_owned()));
+        }
+
+        let cmd = self.cmd.as_ref().unwrap();
+        match cmd {
+            Cmd::Autofill { .. } => self.auto_fill(),
+            Cmd::Dump { .. } => self.dump(),
+            Cmd::Truncate { .. } => self.truncate(),
+            Cmd::Check { path } => self.check(path),
+        }
+    }
+
+    fn auto_fill(&self) -> Result<()> {
+        todo!("unplement!")
```

Comment:
How about we create these interfaces for `Engine` too, and leave their implementation empty?

Reply:
Sounds goods, I will do as your advice.